### PR TITLE
Subscribe to source job worker metrics in job master

### PR DESCRIPTION
**Pull request by calvin681**
### Context

This PR adds support for subscribing to drop metrics from source job workers. The feature is gated by a job property. The source job metrics are funneled to the stage 1 operator since that's the stage you need to scale to mitigate the drops.

The PR lays the groundwork to start receiving source job metrics. There will be followup PRs to actually use these metrics in auto scaling and outlier detection.

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable
- [x] Added copyright headers for new files from `CONTRIBUTING.md`

### CODE REVIEW

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java [hunk 1]
```diff
@@ -280,6 +307,28 @@ public void call() {
...
+                                for (Map.Entry<String, GaugeData> group : metricGroups.entrySet()) {
+                                    String metricKey = worker.getKey() + ":" + group.getKey();
+                                    Map<String, Double> gauges = group.getValue().getGauges();
+                                    if (sourceMetricsRecent.containsKey(metricKey) && gauges.containsKey(DROPPED_COUNTER_METRIC_NAME)) {
+                                        sourceJobDrops += gauges.get(DROPPED_COUNTER_METRIC_NAME);
+                                        hasSourceJobDropsMetric = true;
+                                    }
+                                }
+                            }
+                            if (hasSourceJobDropsMetric) {
```
Comment (neerajrj) [inline mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:325]: Is there a way to weed out outliers? As in say source job is dropping data while writing to to only 1 worker, then we don't necessarily want to scale up this job right

Comment (calvin681) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:325]: Yes, that's planned in a followup PR. The outlier detector should automatically resubmit the worker when it detects drops are only to 1 worker.

File: mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+import io.mantisrx.shaded.com.fasterxml.jackson.annotation.JsonProperty;
+import io.mantisrx.shaded.com.fasterxml.jackson.core.type.TypeReference;
+import io.mantisrx.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public class SourceJobParameters {
+    public static final String MANTIS_SOURCEJOB_TARGET_KEY = "target";
```
Comment (nickmahilani) [inline mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java:31]: This assumes a multi source job connector. For single source job connectors, we would look at the sourceJobName job param.
For MRE v2 app stream connector, there is no source job name in the job params.

We should probably make this more generic for 
1) Source job name config: either identify the source job names that need to be queried (using another job param value or app stream etc) or provide the source job names in the autoscaling config 
2) the metric group patterns to match against

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java [hunk 3]
```diff
@@ -98,6 +120,16 @@ public AggregationAlgo getAggregationAlgo(final String metricGroupName, final St
         return getAllMetrics().keySet();
     }
 
+    public Set<String> generateSourceJobMetricGroups(Set<String> clientIds) {
+        Set<String> results = new HashSet<>();
+        for (String clientId : clientIds) {
+            for (String metricPattern : sourceJobMetrics.keySet()) {
+                results.add(metricPattern.replaceAll(CLIENT_ID_TOKEN, clientId));
```
Comment (nickmahilani) [inline mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java:127]: We could probably just look for all ServerSentEventRequestHandler dropped metrics with current-job-Id(should be in the clientId by default?) in the metric group name.. There will also be sockAddr in the metricGroupName, so there will be multiple metric groups per clientId anyways.

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java [hunk 4]
```diff
@@ -280,6 +307,28 @@ public void call() {
...
+                                logger.info("Job stage " + stage + ", recent source job metrics count: " +
+                                        sourceMetricsRecent.size());
+                                jobAutoScaleObserver.onNext(
+                                        new JobAutoScaler.Event(StageScalingPolicy.ScalingReason.SourceJobDrop, stage,
+                                                sourceJobDrops / 6.0 / numWorkers, numWorkers, ""));
```
Comment (nickmahilani) [inline mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:330]: why 6.0?

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java [hunk 5]
```diff
@@ -280,6 +307,28 @@ public void call() {
...
+
+                            double sourceJobDrops = 0;
+                            boolean hasSourceJobDropsMetric = false;
+                            Map<String, String> sourceMetricsRecent = sourceJobMetricsRecent.asMap();
+                            for (Map.Entry<String, WorkerMetrics> worker : sourceJobWorkersMap.entrySet()) {
+                                Map<String, GaugeData> metricGroups = metricAggregator.getAggregates(worker.getValue().getGaugesByMetricGrp());
+                                for (Map.Entry<String, GaugeData> group : metricGroups.entrySet()) {
+                                    String metricKey = worker.getKey() + ":" + group.getKey();
+                                    Map<String, Double> gauges = group.getValue().getGauges();
+                                    if (sourceMetricsRecent.containsKey(metricKey) && gauges.containsKey(DROPPED_COUNTER_METRIC_NAME)) {
```
Comment (nickmahilani) [inline mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:319]: gauges list won't have the Dropped counter, so this condition will be false always. There is a separate counters list in the metric event. Though, the step based autoscaler has only looked at gauges so far, so the aggregations of the metric data for a single worker and across workers might have some assumptions based on that.

File: mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java (hunk 2 above)
Comment (calvin681) [reply mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java:31]: (1) makes sense. The various connectors may use different params for both `sourceJobName` and `clientId`. There is no clean way to support all of them. I will add a system param called `mantis.jobmaster.autoscale.sourcejob.target` that can be used as override.

(2) what's the use case? Are there other job-to-job transport outside of `ServerSentEventRequestHandler` that can emit drops?

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java (hunk 3 above)
Comment (calvin681) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java:127]: The job can have multiple clientIds. The patterns generated would be `ServerSentEventRequestHandler:clientId=clientId1:*`, etc. `AutoScaleMetricConfigTest` should make this clear. Is this what you mean?

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java (hunk 4 above)
Comment (calvin681) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:330]: Similar to what's already done on the RPS metric: https://github.com/Netflix/mantis/pull/82/files/e489031772bf021ee83fe0c145e1e202dbc0638a#diff-8068879b4809e7e6458c38f3ab1ce02fb794d943a9b0b4a1953d21d1a07948a4R307

This is to get a per second number and relates to how atlas cumulates metrics for 6 seconds.

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java (hunk 5 above)
Comment (calvin681) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:319]: The metric gets converted to gauges here: https://github.com/Netflix/mantis/pull/82/files/e489031772bf021ee83fe0c145e1e202dbc0638a#diff-fce5244027e72de847d3bffc7dfdd658a28fd5ab90b802b0dd662268a1112f2bR100

This is to simplify implementation and similar to how the stage ON_NEXT_GAUGE metric is generated. The other option is to make `MetricData`, `WorkerMetric` classes counter aware. That would be a bigger change.

File: mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java (hunk 2 above)
Comment (nickmahilani) [reply mantis-runtime/src/main/java/io/mantisrx/runtime/parameter/SourceJobParameters.java:31]: 2) no current use case, this is more forward looking if we add a websocket/grpc sink in future.

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java (hunk 5 above)
Comment (nickmahilani) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:319]: oh ok, I missed the counter to gauge conversion, so was confused reading the code.

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java (hunk 4 above)
Comment (nickmahilani) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/WorkerMetricHandler.java:330]: hmm didn't know that the atlas counters are reset every 6 sec.. should probably add a comment

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java (hunk 3 above)
Comment (nickmahilani) [reply mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java:127]: nevermind, I was thinking the `clientId` would always have the current jobID in it, so we could just look for `ServerSentEventRequestHandler:.*clientId=<currentJobId>.*` but it looks like `clientId` can be overridden in the TargetInfo job param.

File: mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java [hunk 6]
```diff
@@ -17,18 +17,28 @@
...
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Pattern;
 
 
 public class AutoScaleMetricsConfig {
 
+    public static final String CLIENT_ID_TOKEN = "_CLIENT_ID_";
+    public static final String OUTBOUND_METRIC_GROUP_PATTERN = "PushServerSse:clientId=" + CLIENT_ID_TOKEN + ":*";
```
Comment (nickmahilani) [inline mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/jobmaster/AutoScaleMetricsConfig.java:32]: suggest using a string constant for `PushServerSse` / `ServerSentEventRequestHandler` and `clientId` that is used for both metric generation and consumption here. So we don't unintentionally break this if/when the metric name is changed or a new tag is added to the metric etc.

**Review by nickmahilani — APPROVED** [review]
LGTM

**Pull request merged**